pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,39 +1244,146 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
 		return "", err
 	}
 
-	// vmdks will be created inside kubevols directory
-	kubeVolsPath := filepath.Clean(ds.Path(VolDir)) + "/"
-	err = makeDirectoryInDatastore(vs.client, dc, kubeVolsPath, false)
-	if err != nil && err != ErrFileAlreadyExist {
-		glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
-		return "", err
+	if volumeOptions.StorageProfileData != "" {
+		// Check if the datastore is VSAN if any capability requirements are specified.
+		// VSphere cloud provider now only supports VSAN capabilities requirements
+		ok, err := checkIfDatastoreTypeIsVSAN(vs.client, ds)
+		if err != nil {
+			return "", fmt.Errorf("Failed while determining whether the datastore: %q"+
+				" is VSAN or not.", datastore)
+		}
+		if !ok {
+			return "", fmt.Errorf("The specified datastore: %q is not a VSAN datastore."+
+				" The policy parameters will work only with VSAN Datastore."+
+				" So, please specify a valid VSAN datastore", datastore)
+		}
 	}
-	glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
 
-	vmDiskPath := kubeVolsPath + volumeOptions.Name + ".vmdk"
+	dummyVMName := "dummy-Kubernetes-vm"
+	virtualMachineConfigSpec := types.VirtualMachineConfigSpec{
```

Comment:
The idea over here is the disk should be created even when no storage profile data is specified. For example consider the case,

kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: slow
provisioner: kubernetes.io/vsphere-volume
parameters:
  diskformat: thin

For the above example, disk should be created even when there is no storage profile data. How I would be doing this create dummy VM with disk attached to it no profile configured to the disk. You can see in the same function where I make a check to configure policy to disk if storageProfile is not null. 

Previously we are directly creating the disk on VM directly. I have removed this logic completely. Now in both cases we would create dummy VM with disk . But disk will be configured policy or not depends on volumeOptions.StorageProfileData.